include/seastar/core/gate.hh

```diff
@@ -47,21 +51,36 @@ public:
 class gate {
     size_t _count = 0;
     std::optional<promise<>> _stopped;
+
+#ifdef GATE_HOLDER_DEBUG
+    size_t _holders = 0;
+
+    void assert_not_held(const gate& g) const noexcept {
+        assert(!g._holders && "gate moved with outstanding holders");
+    }
+#else   // GATE_HOLDER_DEBUG
+    void assert_not_held(const gate& g) const noexcept {}
```

Comment:
so we don't need to worry about a picky compiler which might warn on a defined-but-not used variable.